tests/test_preprocessing_utils.py

```diff
@@ -180,9 +191,581 @@ def test_toy_dataset(self):
+        ds = ds.map(lambda ex: datasource_preprocessor.preprocess(ex), batched=True, batch_size=3)
+
+        self.assertEqual(ds[:][col_to_store_text], self.target_texts)
+        self.assertEqual(ds[:][col_to_store_metadata_html], self.target_metadata_html)
+        self.assertEqual(ds[:][col_to_store_metadata_url], self.target_metadata_url)
+        self.assertEqual(ds[:][col_to_store_metadata_timestamp], self.target_metadata_timestamp)
+        self.assertEqual(ds[:][col_to_store_metadata_website_desc], self.target_metadata_website_desc)
+        self.assertEqual(ds[:][col_to_store_metadata_entities], self.target_metadata_entities)
+        self.assertEqual(
+            ds[:][col_to_store_metadata_generation_length_text], self.target_metadata_generation_length_text
+        )
+        self.assertEqual(
+            ds[:][col_to_store_metadata_generation_length_sentence], self.target_metadata_generation_length_sentence
+        )
+        self.assertEqual(ds[:][col_to_store_metadata_datasource], self.target_metadata_datasource)
+
+
+        col_to_store_all_metadata = "metadata"
+        columns_names_to_concat = [
+            col_to_store_metadata_html,
+            col_to_store_metadata_url,
+            col_to_store_metadata_timestamp,
+            col_to_store_metadata_website_desc,
+            col_to_store_metadata_entities,
+            col_to_store_metadata_generation_length_text,
+            col_to_store_metadata_generation_length_sentence,
+            col_to_store_metadata_datasource,
+        ]
+        concat_columns_fn = concat_columns(
+            columns_names_to_concat=columns_names_to_concat,
+            new_colum_name=col_to_store_all_metadata,
+        )
+        ds = ds.map(concat_columns_fn, batched=True, batch_size=3, remove_columns=columns_names_to_concat)
+
+        for metadata_type in [
+            self.target_metadata_html,
+            self.target_metadata_url,
+            self.target_metadata_timestamp,
+            self.target_metadata_website_desc,
+            self.target_metadata_entities,
+            self.target_metadata_generation_length_text,
+            self.target_metadata_generation_length_sentence,
+            self.target_metadata_datasource,
+        ]:
+            for id, metadata_example in enumerate(metadata_type):
+                for metadata in metadata_example:
+                    for potential_missing_key in [
+                        "char_end_idx",
+                        "char_start_idx",
+                        "relative_end_pos",
+                        "relative_start_pos",
+                        "html_attrs",
+                        "ent_desc",
+                    ]:
+                        if potential_missing_key in metadata:
+                            continue
+                        metadata[potential_missing_key] = None
+                    self.assertIn(metadata, ds[id][col_to_store_all_metadata])
```

Comment:
@cccntu , to accompany my answer on slack, here is the part I just added to the tests to show how we can reconcile all the metadata in the same column at the end